src/graph.hpp

```diff
@@ -2,15 +2,46 @@
 #define GRAPH_HPP
 
 #include <edge.hpp>
+#include <impl/graph.hpp>
+#include <impl/meta_utils.hpp>
 
 namespace graph {
 
-template<typename Node, typename... Types>
-class Graph {
+class GraphTrait {
+protected:
+    GraphTrait() = default;
+};
+
+class EdgeTrait {
+protected:
+    EdgeTrait() = default;
+};
+
+// TODO: source and sink Node* type.
+// Should we inherit all nodes from base Node class
+// or should we apply more template woodoo?
+class Net : public GraphTrait {
```

Comment:
Make `Net` a template class.

Reply:
We have unpleasant interface for user in this case
```C++
graph::Graph<int, graph::Net<int>, graph::Weighted> g;
```
```graph::Net<int>``` user have to write ```int``` second time here (obviously it's unnecessary) and can make mistake

Reply:
It should result in a compile-time error. It is normal, people make mistakes.

Reply:
Probably ```void*``` with ```reinterpret_cast``` will work here

Reply:
It seems to me that is a bad idea to use void* in this case. You say:

> user have to write int second time here (obviously it's unnecessary) and can make mistake

, but at the same time suggest use `reinterpret_cast` with `void*`. Don't throw your code away, son. 😄